[RFC][mlir] ViewLikeOpInterface method for detecting partial views.

vzakhari opened the pull request:
I am trying to simplify the implementation of Flang's AliasAnalysis
by using ViewLikeOpInterface for some FIR dialect operations.
Currently, Flang's AliasAnalysis implementation explicitly recognizes
specific set of operations, which is inconvenient.
ViewLikeOpInterface may also be used in other parts of Flang.

This patch adds isKnownToHaveSameStart() method to detect
cases where the input and resulting views may be different
by an offset, in which case FIR alias analysis must not return
`MustAlias`.

New method isKnownToBeCompleteView() returns true iff
the resulting view is a complete view of the input view.
This method will be used by the following patches for OpenACC
support in Flang.

This patch also modifies LocalAliasAnalysis to stop returning
`MustAlias` in cases where both `MustAlias` and `PartialAlias`
are possible - we will return `MayAlias` now, which is conservatively
correct.

jeanPerier reviewed (commented):
It makes sense to me to add the ViewLikeOpInterface to these FIR operations, thanks Slava.

> If my understanding of PartialAlias is correct, then we can have a partial alias even when the starting addresses of the source and the resulting view are the same - though, I might be wrong.

That is also my reading of LLVM/MLIR documentation, although I have little experience with how it is used in practice.

tblah commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -212,6 +212,10 @@ AliasResult AliasAnalysis::alias(Source lhsSrc, Source rhsSrc, mlir::Value lhs,
+      // TODO: we should return PartialAlias here. Need to decide
+      // if returning PartialAlias for fir.pack_array is okay;
+      // if not, then we need to handle it specially to still return
+      // MayAlias.
```
From the LLVM documentation:
> The PartialAlias response is used when the two memory objects are _known to be overlapping_ in some way, regardless of whether they start at the same address or not.

I guess it depends what exactly is meant by "object". Previously I had interpreted it as referring to the specific memory addresses which could be accessed. So `array(1:4)` and `array(3)` do overlap with each other but not with `array(100:200)`. Therefore I think array indexing (until analysed more carefully) should use `MayAlias` rather than `PartialAlias`.

Alternatively, if "object" means the entire allocation then `PartialAlias` makes sense here.

vzakhari replied:
I agree with you, Tom. I think the current algorithm in LocalAliasAnalysis (which I am not trying to affect too much, because this is not my goal) does not allow reasoning about partial aliases quite precisely.  So I feel okay returning `MayAlias` in cases where there might be an offset involved. At least, this makes the analysis conservatively correct.

tblah reviewed (commented):
Thanks Slava. Really interesting work.

vzakhari committed:
[RFC][mlir] ViewLikeOpInterface method for detecting partial views.

I am trying to simplify the implementation of Flang's AliasAnalysis
by using ViewLikeOpInterface for some FIR dialect operations.
Currently, Flang's AliasAnalysis implementation explicitly recognizes
specific set of operations, which is inconvenient.
ViewLikeOpInterface may also be used in other parts of Flang.

This patch adds isKnownToHaveSameStart() method to detect
cases where the input and resulting views may be different
by an offset, in which case FIR alias analysis must not return
`MustAlias`.

New method isKnownToBeCompleteView() returns true iff
the resulting view is a complete view of the input view.
This method will be used by the following patches for OpenACC
support in Flang.

This patch also modifies LocalAliasAnalysis to stop returning
`MustAlias` in cases where both `MustAlias` and `PartialAlias`
are possible - we will return `MayAlias` now, which is conservatively
correct.

vzakhari committed:
Renamed the methods to clearly indicate what their result means.

razvanlupusoru reviewed (approved):
With the default conservative behavior of returning false of `isKnownToBeCompleteView`, I worry it is "incomplete" (for my own needs not for Alias Analysis). The missing information that I was hoping to rely on is the property of `isKnownToBePartialView`. I had assumed previously with the addition of `isCompleteView` that returning false would mean I could make that assumption.

FWIW though I am comfortable with the current changes and thus approving to let the maintainers know you have additional support because I believe this is heading in the right direction to hold additional information critical for AA. Thank you!

matthias-springer commented on mlir/include/mlir/Interfaces/ViewLikeInterface.td:
```diff
@@ -23,21 +23,53 @@ def ViewLikeOpInterface : OpInterface<"ViewLikeOpInterface"> {
+             are known to start at the same "address".
+           }],
+           /*retTy=*/"bool",
+           /*methodName=*/"isKnownToHaveSameStart",
```
Can `isKnownToHaveSameStart` and `isKnownToBeCompleteView` be replaced by a single interface method that returns `AliasResult::Kind`?

vzakhari replied:
Yes, I think adding a method like `alias` will allow me to make the same changes in the alias analyses.

At the same time, I think the new `alias` method may become redundant if we decide to add the two more "basic" methods like `isKnownToHaveSameStart`/`isKnownToBeCompleteView` in future.

I think the following pieces of "basic" information provide everything we need to make a decision about the aliasing of the input and resulting views:
* Can we prove that the resulting view is a complete view of the input view? This method may return `true` even in cases when the views have dynamic sizes, e.g. based on the semantics of the concrete operation.
* Constant size of the input and the resulting views (or nullopt).
* Constant offset applied to the resulting input view to produce the resulting view (or nullopt).  `isKnownToHaveSameStart` is a simplified version of this query, i.e. it only says if the offset is zero and does not care to compute the actual non-zero constant offset, if it is possible.

I believe having these "basic" methods available, we can implement `alias` for all operations with `ViewLikeOpInterface`, and the concrete operations won't have any extra knowledge that will make sense overriding the `alias` method. That is why I decided to go with more "basic" methods (that may also be useful outside of the alias analysis, probably).

With that said, I agree replacing the two methods with an `alias` method. We can always adjust it in future if we end up adding all the method providing the "basic" information.

I can think of the following specification for the new `alias` method:
* This method returns AliasResult::Kind representing the aliasing between the input and the resulting views of a `ViewLikeOpInterface` operation.
* It returns `NoAlias` only when the resulting view is known to have zero size.  Note that when the input view has zero size the resulting view must have zero size as well. Otherwise,
* it returns `MustAlias` when the resulting view is known to start at the same address as the input view. Otherwise,
* it returns `PartialView` when the resulting view is known to start at a non-zero offset of the input view. Otherwise,
* it returns `MayAlias`.

Does this definition make sense to you?

vzakhari replied:
Matthias and I discussed this offline. Matthias raised a valid concern that it is unclear what the aliasing rules are for the memref subviews, as long as the subview actually holds the original base pointer of the view and the offset. Should the aliasing replies be based on the base addresses or their offsetted values - this seems to require a broader discussion in MLIR community.

We agreed that I implement a Flang specific interface to help making FIR alias analysis more generic, and then we'll present the working implementation in an RFC and discuss whether it is worth making this interface MLIR-global.

I will close this PR and will work on the new interface in Flang.

matthias-springer commented on flang/lib/Optimizer/Analysis/AliasAnalysis.cpp:
```diff
@@ -559,10 +563,19 @@ AliasAnalysis::Source AliasAnalysis::getSource(mlir::Value v,
-        .Case<fir::ConvertOp>([&](auto op) {
-          // Skip ConvertOp's and track further through the operand.
-          v = op->getOperand(0);
+        .Case<mlir::ViewLikeOpInterface>([&](auto op) {
```
Why does Flang have its own alias analysis? How is it different from the MLIR alias analysis? Given that you are improving this interface, could the Flang alias analysis be replaced with the one from MLIR?

vzakhari replied:
Flang alias analysis has special handling for HLFIR/FIR operations that carry extra information about the rules of aliasing for objects created from Fortran variables. For example, there are special rules for DUMMY arguments of functions that cannot alias each other unless they have TARGET or POINTER attributes. Flang carries this extra information on `[hl]fir.declare` operations. For the two given "addresses", Flang alias analysis tries to discover the "sources" of those addresses, e.g. by tracking through view-like operations, and then reasons about the aliasing having reached `[hl]fir.declare` operations.

I think the HLFIR/FIR specifics used by Flang alias analysis can be abstracted by using some generic interfaces that `[hl]fir.declare` (and some other operations, e.g. `fir.global`), but I want to leave this out of scope of this PR.

matthias-springer commented on mlir/include/mlir/Analysis/AliasAnalysis.h:
```diff
@@ -35,9 +35,12 @@ class AliasResult {
     MayAlias,
-    /// The two locations alias, but only due to a partial overlap.
+    /// The two locations overlap in some way, regardless of whether
+    /// they start at the same address or not.
```
This seems to change the semantics of `PartialAlias`. Is it safe?

matthias-springer commented on mlir/include/mlir/Analysis/AliasAnalysis.h:
```diff
@@ -35,9 +35,12 @@ class AliasResult {
+    /// The two locations precisely alias each other, meaning that
+    /// they always start at exactly the same location.
+    /// This result does not imply that the pointers compare equal.
     MustAlias,
```
I think it's worth mentioning as a comment of the enum that these values are taken from LLVM.

matthias-springer commented on mlir/lib/Analysis/AliasAnalysis/LocalAliasAnalysis.cpp:
```diff
@@ -40,17 +40,19 @@ using namespace mlir;
 static constexpr unsigned maxUnderlyingValueSearchDepth = 10;
 
 /// Given a value, collect all of the underlying values being addressed.
```
Can you add documentation for the arguments? Here and for all functions who's signature has changed.

matthias-springer commented on mlir/lib/Analysis/AliasAnalysis/LocalAliasAnalysis.cpp:
```diff
@@ -40,17 +40,19 @@ using namespace mlir;
+static void
+collectUnderlyingAddressValues(Value value, unsigned maxDepth,
+                               DenseMap<Value, bool> &visited, bool maybeOffset,
+                               SmallVectorImpl<std::pair<Value, bool>> &output);
```
I'm not familiar this function. Why is this change needed?
